[mac] WebDriver sometimes taking screenshots with transparent grey line at the top and no rounded corners

the-chenergy opened the pull request:
#### bdd0eaf1d28f834bdc3fab3e656486522f3b301f
<pre>
[mac] WebDriver sometimes taking screenshots with transparent grey line at the top and no rounded corners
<a href="https://rdar.apple.com/116020785">rdar://116020785</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=262078">https://bugs.webkit.org/show_bug.cgi?id=262078</a>

Reviewed by Aditya Keerthi, Abrar Rahman Protyasha and BJ Burg.

Our WebDriver currently takes viewport screenshots with the
`static takeWindowSnapshot` function in WebViewImpl. What that function
currently does is:

   1. Try taking a screenshot using the hardware capture path, with the
      CGSHWCaptureWindowList function and the
      kCGSCaptureIgnoreGlobalClipShape option to avoid the rounded
      corners.
   2. The above function may fail (e.g. due to the window being off-
      screen, as noted in the code comments), and if it does, take
      a screenshot using the non-hardware capture path with
      CGWindowListCreateImage and kCGWindowImageBoundsIgnoreFraming.

Due to the kCGSCaptureIgnoreGlobalClipShape option currently bugged
and having no effect, if hardware capturing does succeed, we get a
screenshot with the rounded corners, and if it fails we get no
rounded corners, which contributes toward the tests&apos; flakiness.

   - This inconsistency is also observable from the screenshots taken
     when performing a swipe to go back/forward in the browser, though
     the consequence there is less severe than failing WPT tests.

To address that, we force WebDriver to always fall back to the non-
hardware capture path, taking viewport screenshots with
CGWindowListCreateImage, to rule out any flakiness caused by
CGSHWCaptureWindowList occasionally failing.

   - The reason we began to prefer the hardware capture path with
     CGSHWCaptureWindowList is because CGWindowListCreateImage had a
     performance issue where it might hang. Source: <a href="https://github.com/WebKit/WebKit/commit/8f81edf40d069cacf6107254a535a812d5978dea">https://github.com/WebKit/WebKit/commit/8f81edf40d069cacf6107254a535a812d5978dea</a>
   - However, the hang has been addressed internally and no longer
     occurs today. My local testing instead shows no noticeable
     difference in speed between CGWindowListCreateImage and
     CGSHWCaptureWindowList.
   - Though, CGWindowListCreateImage has been deprecated in favor of the
     ScreenCaptureKit, but it&apos;s still functional for now. So, I believe
     this patch is a reasonable workaround at least until the hardware
     capturing bug gets fixed, in exchange for a more consistent
     performance in WPT.

Despite this patch aims to improve WPT results, it&apos;s hard to pinpoint
which specific tests will progress, as all tests that take viewport
screenshots are currently flaky due to the bug. However, I can observe
some tests progressing locally. With this command inside the WPT repo:

    ./wpt run safari ./css/css-contain --webdriver-binary=&quot;...&quot;

556 tests were run, where 140-150 tests fail without this patch, and
this patch makes it so that exactly 51 tests were still failing (about
90 tests progressed) and no tests regressed.

* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::takeViewSnapshot):
   - Add the option to always use the non-hardware screen capture path
     when taking a window snapshot.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::takeScreenshot):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/ViewSnapshotStore.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::takeViewSnapshot):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::takeViewSnapshot):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::takeWindowSnapshot):
   - Allow WebDriver to make use of this change when taking a viewport
     screenshot.

Canonical link: <a href="https://commits.webkit.org/281887@main">https://commits.webkit.org/281887@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b838dd5bf845a6bd4f6ae3a6d913859cd40acd63

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61262 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40623 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13843 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65212 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11811 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48301 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12086 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49506 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8210 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/63296 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37778 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53079 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30338 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34455 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10724 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56272 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10599 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66943 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5209 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10396 "Found 1 new test failure: fast/forms/form-submission-crash-4.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56879 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5232 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53042 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57101 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13671 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4304 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/36427 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37510 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38604 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37254 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -2400,7 +2400,8 @@ void WebAutomationSession::takeScreenshot(const Inspector::Protocol::Automation:
     Function<void(WebPageProxy&, std::optional<WebCore::IntRect>&&, Ref<TakeScreenshotCallback>&&)> takeViewSnapsot = [](WebPageProxy& page, std::optional<WebCore::IntRect>&& rect, Ref<TakeScreenshotCallback>&& callback) {
         page.callAfterNextPresentationUpdate([page = Ref { page }, rect = WTFMove(rect), callback = WTFMove(callback)] () mutable {
-            auto snapshot = page->takeViewSnapshot(WTFMove(rect));
+            const bool alwaysUseNonHardwareCapturePath = true;
```
We would normally make `takeViewSnapshot` take an enum class parameter with `::Yes`/`::No`, rather than introduce a `const bool` here.

pxlcoder commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -2400,7 +2400,8 @@ void WebAutomationSession::takeScreenshot(const Inspector::Protocol::Automation:
         page.callAfterNextPresentationUpdate([page = Ref { page }, rect = WTFMove(rect), callback = WTFMove(callback)] () mutable {
-            auto snapshot = page->takeViewSnapshot(WTFMove(rect));
+            const bool alwaysUseNonHardwareCapturePath = true;
+            RefPtr<ViewSnapshot> snapshot = page->takeViewSnapshot(WTFMove(rect), alwaysUseNonHardwareCapturePath);
```
```suggestion
            RefPtr snapshot = page->takeViewSnapshot(WTFMove(rect), alwaysUseNonHardwareCapturePath);
```

pxlcoder commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -4504,18 +4504,28 @@ static NSPasteboardName pasteboardNameForAccessCategory(WebCore::DOMPasteAccessC
+    // the kCGSCaptureIgnoreGlobalClipShape option has no effect and the resulting screenshot
+    // still contains the window's rounded corners. There are WPT tests relying on comparing
+    // web driver's screenshots that cannot tolerate this inconsistency, especially due to
+    // CGSHWCaptureWindowList not always succeeding. So for web driver only, we bypass that bug
```
```suggestion
    // CGSHWCaptureWindowList not always succeeding. So for WebDriver only, we bypass that bug
```

pxlcoder commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -4504,18 +4504,28 @@ static NSPasteboardName pasteboardNameForAccessCategory(WebCore::DOMPasteAccessC
+    // FIXME <https://webkit.org/b/??????>: CGSHWCaptureWindowList is currently bugged where
+    // the kCGSCaptureIgnoreGlobalClipShape option has no effect and the resulting screenshot
+    // still contains the window's rounded corners. There are WPT tests relying on comparing
+    // web driver's screenshots that cannot tolerate this inconsistency, especially due to
```
```suggestion
    // WebDriver's screenshots that cannot tolerate this inconsistency, especially due to
```

pxlcoder commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -4504,18 +4504,28 @@ static NSPasteboardName pasteboardNameForAccessCategory(WebCore::DOMPasteAccessC
+        CGSWindowCaptureOptions options = kCGSCaptureIgnoreGlobalClipShape;
+        if (captureAtNominalResolution)
+            options |= kCGSWindowCaptureNominalResolution;
+        RetainPtr<CFArrayRef> windowSnapshotImages = adoptCF(CGSHWCaptureWindowList(CGSMainConnectionID(), &windowID, 1, options));
```
```suggestion
        RetainPtr windowSnapshotImages = adoptCF(CGSHWCaptureWindowList(CGSMainConnectionID(), &windowID, 1, options));
```

the-chenergy replied:
Is this really a good idea to simply? It's not easy to infer the return type of `CGSHWCaptureWindowList` from its function name, so I like that it's explicitly spelled out here.

pxlcoder replied:
It's still usually omitted – but I don't feel strongly about it, so feel free to keep it!

pxlcoder commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -4504,18 +4504,28 @@ static NSPasteboardName pasteboardNameForAccessCategory(WebCore::DOMPasteAccessC
+    }
 
-    // Fall back to the non-hardware capture path if we didn't get a snapshot
-    // (which usually happens if the window is fully off-screen).
```
I don't think it makes sense to move this comment?

burg commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -2400,7 +2400,7 @@ void WebAutomationSession::takeScreenshot(const Inspector::Protocol::Automation:
 #if PLATFORM(GTK) || PLATFORM(COCOA)
     Function<void(WebPageProxy&, std::optional<WebCore::IntRect>&&, Ref<TakeScreenshotCallback>&&)> takeViewSnapsot = [](WebPageProxy& page, std::optional<WebCore::IntRect>&& rect, Ref<TakeScreenshotCallback>&& callback) {
         page.callAfterNextPresentationUpdate([page = Ref { page }, rect = WTFMove(rect), callback = WTFMove(callback)] () mutable {
-            auto snapshot = page->takeViewSnapshot(WTFMove(rect));
```
Nit: leave auto, or add type argument to RefPtr<T>

pxlcoder replied:
`RefPtr` is WebKit style, per explicit smart pointer guidelines.

burg commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -1764,7 +1764,8 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
 
 #if PLATFORM(COCOA) || PLATFORM(GTK)
-    RefPtr<ViewSnapshot> takeViewSnapshot(std::optional<WebCore::IntRect>&&);
+    enum class ShouldAlwaysUseNonHardwareCapturePath : bool { No, Yes };
```
Naming: I think 'ForceSoftwareRendering' or similar is better than 'ShouldAlwaysUseNonHardwareCapturePath'.

burg commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -4504,15 +4504,25 @@ static NSPasteboardName pasteboardNameForAccessCategory(WebCore::DOMPasteAccessC
 }
 
-static RetainPtr<CGImageRef> takeWindowSnapshot(CGSWindowID windowID, bool captureAtNominalResolution)
+static RetainPtr<CGImageRef> takeWindowSnapshot(CGSWindowID windowID, bool captureAtNominalResolution, bool alwaysUseNonHardwareCapturePath)
```
Shouldn't this use the same Yes/No enum as defined in the header?

burg reviewed (commented):
Since the commit message mentions failing tests, please mention which WPT tests are progressing as a result of this change.

aprotyas commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -4504,15 +4504,24 @@ static NSPasteboardName pasteboardNameForAccessCategory(WebCore::DOMPasteAccessC
-    if (captureAtNominalResolution)
-        options |= kCGSWindowCaptureNominalResolution;
-    RetainPtr<CFArrayRef> windowSnapshotImages = adoptCF(CGSHWCaptureWindowList(CGSMainConnectionID(), &windowID, 1, options));
+    // FIXME <https://webkit.org/b/??????>: CGSHWCaptureWindowList is currently bugged where
```
Is there a bug ID for this?

pxlcoder commented on Source/WebKit/UIProcess/mac/PageClientImplMac.h:
```diff
@@ -171,6 +171,7 @@ class PageClientImpl final : public PageClientImplCocoa
     void didFirstLayerFlush(const LayerTreeContext&) override;
 
     RefPtr<ViewSnapshot> takeViewSnapshot(std::optional<WebCore::IntRect>&&) override;
+    RefPtr<ViewSnapshot> takeViewSnapshotAlwaysUsingNonHardwareCapturePath(std::optional<WebCore::IntRect>&&) override;
```
Why is this a separate method, and not reusing the flag we already have?

the-chenergy replied:
I did this to avoid having to change multiple `.h` files that also overrides `takeViewSnapshot` (like `PageClientImplIOS.h`).

the-chenergy replied:
Or I guess we could've also made an override that takes two arguments. Hmm, maybe that's better actually.

the-chenergy commented on Source/WebKit/UIProcess/PageClient.h:
```diff
@@ -195,7 +196,6 @@ class WebContextMenuProxy;
 class WebEditCommandProxy;
 class WebFrameProxy;
 class WebOpenPanelResultListenerProxy;
-class WebPageProxy;
```
I had to do this because this `class WebPageProxy` forward declaration doesn't allow me to use `WebPageProxy::ForceSoftwareCapturing` to access its nested enum.

Aditya, do you know if there was a particular reason we chose to forward declare originally? @pxlcoder

aprotyas replied:
Not having to pull in `WebPageProxy.h` is a plus for build times.

aprotyas replied:
We prefer forward declaring types in headers when we can.

pxlcoder replied:
I think we should not use a nested enum. It's highly preferable to avoid adding an include to WebPageProxy.

the-chenergy replied:
OK. Do you think it'll be better to revert this file back to where I had `takeViewSnapshotAlwaysUsingNonHardwareCapturePath` rather than an overload, keeping the nested enum? Otherwise where are you thinking I should put that enum?

pxlcoder replied:
How about moving the enum class to `ViewSnapshotStore.h`?

pxlcoder commented on Source/WebKit/UIProcess/PageClient.h:
```diff
@@ -195,7 +196,6 @@ class WebContextMenuProxy;
 class WebEditCommandProxy;
 class WebFrameProxy;
 class WebOpenPanelResultListenerProxy;
-class WebPageProxy;
```
Don't think we should be removing this forward declaration.

the-chenergy committed:
[mac] WebDriver sometimes taking screenshots with transparent grey line at the top and no rounded corners
rdar://116020785
https://bugs.webkit.org/show_bug.cgi?id=262078

Reviewed by Aditya Keerthi, Abrar Rahman Protyasha and BJ Burg.

Our WebDriver currently takes viewport screenshots with the
`static takeWindowSnapshot` function in WebViewImpl. What that function
currently does is:

   1. Try taking a screenshot using the hardware capture path, with the
      CGSHWCaptureWindowList function and the
      kCGSCaptureIgnoreGlobalClipShape option to avoid the rounded
      corners.
   2. The above function may fail (e.g. due to the window being off-
      screen, as noted in the code comments), and if it does, take
      a screenshot using the non-hardware capture path with
      CGWindowListCreateImage and kCGWindowImageBoundsIgnoreFraming.

Due to the kCGSCaptureIgnoreGlobalClipShape option currently bugged
and having no effect, if hardware capturing does succeed, we get a
screenshot with the rounded corners, and if it fails we get no
rounded corners, which contributes toward the tests' flakiness.

   - This inconsistency is also observable from the screenshots taken
     when performing a swipe to go back/forward in the browser, though
     the consequence there is less severe than failing WPT tests.

To address that, we force WebDriver to always fall back to the non-
hardware capture path, taking viewport screenshots with
CGWindowListCreateImage, to rule out any flakiness caused by
CGSHWCaptureWindowList occasionally failing.

   - The reason we began to prefer the hardware capture path with
     CGSHWCaptureWindowList is because CGWindowListCreateImage had a
     performance issue where it might hang. Source: https://github.com/WebKit/WebKit/commit/8f81edf40d069cacf6107254a535a812d5978dea
   - However, the hang has been addressed internally and no longer
     occurs today. My local testing instead shows no noticeable
     difference in speed between CGWindowListCreateImage and
     CGSHWCaptureWindowList.
   - Though, CGWindowListCreateImage has been deprecated in favor of the
     ScreenCaptureKit, but it's still functional for now. So, I believe
     this patch is a reasonable workaround at least until the hardware
     capturing bug gets fixed, in exchange for a more consistent
     performance in WPT.

Despite this patch aims to improve WPT results, it's hard to pinpoint
which specific tests will progress, as all tests that take viewport
screenshots are currently flaky due to the bug. However, I can observe
some tests progressing locally. With this command inside the WPT repo:

    ./wpt run safari ./css/css-contain --webdriver-binary="..."

556 tests were run, where 140-150 tests fail without this patch, and
this patch makes it so that exactly 51 tests were still failing (about
90 tests progressed) and no tests regressed.

* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::takeViewSnapshot):
   - Add the option to always use the non-hardware screen capture path
     when taking a window snapshot.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::takeScreenshot):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/ViewSnapshotStore.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::takeViewSnapshot):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::takeViewSnapshot):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::takeWindowSnapshot):
   - Allow WebDriver to make use of this change when taking a viewport
     screenshot.

Canonical link: https://commits.webkit.org/281887@main
